clients/pkg/promtail/targets/gcplog/pull_target.go

```diff
@@ -0,0 +1,164 @@
+package gcplog
+
+import (
+	"context"
+	"sync"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/relabel"
+
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/clients/pkg/promtail/scrapeconfig"
+	"github.com/grafana/loki/clients/pkg/promtail/targets/target"
+)
+
+// PullTarget represents the target specific to GCP project, with a pull subscription type.
+// It collects logs from GCP and push it to Loki.
+// nolint:revive
+type PullTarget struct {
+	metrics       *Metrics
+	logger        log.Logger
+	handler       api.EntryHandler
+	config        *scrapeconfig.GCPLogTargetConfig
+	relabelConfig []*relabel.Config
+	jobName       string
+
+	// lifecycle management
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	// pubsub
+	ps   *pubsub.Client
+	msgs chan *pubsub.Message
+}
+
+// newPullTarget returns the new instance of PullTarget for
+// the given `project-id`. It scraps logs from the GCP project
+// and push it Loki via given `api.EntryHandler.`
+// It starts the `run` loop to consume log entries that can be
+// stopped via `target.Stop()`
+// nolint:revive,govet
+func newPullTarget(
+	metrics *Metrics,
+	logger log.Logger,
+	handler api.EntryHandler,
+	relabel []*relabel.Config,
+	jobName string,
+	config *scrapeconfig.GCPLogTargetConfig,
+) (*PullTarget, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	ps, err := pubsub.NewClient(ctx, config.ProjectID)
+	if err != nil {
+		return nil, err
+	}
+
+	target := newGcplogTarget(metrics, logger, handler, relabel, jobName, config, ps, ctx, cancel)
+
+	go func() {
+		_ = target.run()
+	}()
+
+	return target, nil
+}
+
+// nolint:revive
+func newGcplogTarget(
```

Comment:
nit: maybe this one should also be renamed?

Reply:
this is just returning the struct. why don't we make it as part of `newPullTarget` itself?. Any reason to have it as separate function?